brave/src/test/java/com/linecorp/armeria/server/brave/ArmeriaHttpServerAdapterTest.java

```diff
@@ -165,4 +165,12 @@ public void route_prefix() {
         when(context.route()).thenReturn(Route.builder().path("exact:/foo").build());
         assertThat(ArmeriaHttpServerAdapter.get().route(requestLog)).isEqualTo("/foo");
     }
+
+    @Test
+    public void route_pathWithPrefix() {
+        final ServiceRequestContext context = mock(ServiceRequestContext.class);
+        when(requestLog.context()).thenReturn(context);
+        when(context.route()).thenReturn(Route.builder().pathWithPrefix("/foo/", "glob:bar").build());
+        assertThat(ArmeriaHttpServerAdapter.get().route(requestLog)).isEqualTo("/foo/ ^/(?:.+/)?bar$");
```

Comment:
Why is this not concatenated? a space here what does this help qualify?

ex this will be `get /foo/ ^/(?:.+/)?bar$` for what seems to be much simpler as `get /foo/**bar$`

what am I missing?

Reply:
Because a regular expression is complex and we can never know how to simplify an arbitrary regular expression into a glob pattern?

Reply:
I guess the declaration of the route expression with literal `glob:bar` is lost then? it would be much easier to read if it weren't..

Reply:
main idea.. and I understand this may not happen here.. is that there is parity between the definition and the result. ex if the definition is a simple prefix with glob, what the user put in could be in the http.route tag. you can consider this not binding comment just I want armeria to appear best!

Reply:
Doesn't have strong opinion here, I think it's for the case that we write like
`sb.annotatedService("/6", new MyAnnotatedService6())` with path setting like `"regex:^/regex/(?<path>.*)$"` or `glob:/glob1/*`. 

https://github.com/line/armeria/blob/master/core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceTest.java#L117

If we want to enhance it, we may need create new method for `PathMapping`?

Reply:
I think metertag is just an implementation choice of getting the http route template. This is one good thing about package private as we can change it :)

fwiw there are other things that have regex even if path templates are more common. ex @shimamoto maintains play zipkin tracing which has some cleaning to make it look less like cat walked on keyboard https://github.com/openzipkin-contrib/play-zipkin-tracing/blob/master/play-zipkin-tracing/play/src/main/scala/brave/play/filter/ZipkinTraceFilter.scala#L53

Reply:
What do you mean package private ?
Let me check if I can replace the pattern to ** or * in all cases. WDYT? @trustin

Reply:
And it's better to create new method and pre-calculate the route string instead of always `replace` in `adapter#route`, right?

Reply:
> Let me check if I can replace the pattern to ** or * in all cases. WDYT?

Wouldn't it cause lots of ambiguity when a user regex/glob mapping extensively? I still prefer leaving the expression as is or doing minimal clean-up only for reduced confusion.

Reply:
> I guess the declaration of the route expression with literal glob:bar is lost then? it would be much easier to read if it weren't..

Ah yeah, that's true. I totally missed that the regular expression was derived from a glob expression. Sorry!

Anyways, I think that should be solved by fixing our routing implementation, like adding 'prefix + glob' type.

Reply:
Actually, for prefix + glob case, it could be simplified into a single glob expression. Here:

https://github.com/line/armeria/blob/f7e25f9081d41008443bd0896d8a28d1e4e4ca95/core/src/main/java/com/linecorp/armeria/server/RouteBuilder.java#L194-L199

we could prepend `**/` when the glob expression does not start with `/`, e.g. (prefix = `/foo/` and glob = `bar`) could be simplified into `/foo/**/bar`.

Reply:
@adriancole, This is what you meant, right? :sweat_smile: 

So, my suggestion is:

- for the case of (prefix + regex), leave the regex as is for less ambiguity.
  - I'm not completely sure `prefix + ' ' + regex` is the best. Any suggestions? Does `'(' + prefix + ", " + regex + ')'` look better?
  - +1 for pre-calculating but we need to discuss about where the precalculated string would be stored, API design, etc.
- for the case of (prefix + glob), let us fix the bug where it is not simplified into a glob, which is not the scope of this pull request.

WDYT?

Reply:
I agree that simplifying like the `/foo/**/bar` will be far more readable.

wrt nesting, all the other frameworks with nested expressions join on '/'. I understand there is certain nuance about what happens with multiple expressions that can yield the same pattern. Moreover, sometimes a request is internally routed multiple times. I do understand that there is some complexity some people want to see somewhere. However, I really do think joining on not slash is the most distracting choice to make as an input to the span name. If you feel strongly about doing space joining, it is ok, just mentioning this would be the only framework doing this. It seems like if someone really needed to know the exact difference between these cases, they could add a tag and/or the controller method.

anyway I think we've spent plenty of think time on this, I appreciate you hearing it out. I think you have the understanding needed to make a good call and both choices are moot for those who avoid this pathWithPrefix method.

Reply:
+1 for `/foo/**/bar`.
And for the case of (prefix + regex), I think `'(' + prefix + ", " + regex + ')'` looks better.

Reply:
@adriancole If you are strong about this, then I'm OK with joining on `/`. It is at least better than joining without any separator. :sweat_smile:

Reply:
So when the prefix is `/foo/` and the regex is `^/(?:.+/)?bar$`,
will it be `/foo//^/(?:.+/)?bar$`

Reply:
Because the `prefix` already ends with `/`, we can just concat without any separator, like before.

Reply:
mild request is to strip all redundancy here `//^/` otherwise let's get good enough merged :P

Reply:
Yes, it's okay for me. :-)
Thanks for pointing this out anyway. I was feeling somehow the meter tag is unclear and needed to be fixed when I was working on it. But I didn't do it yet. 😄